File: 11-valid-anagram/solution.py
# Valid Anagram
# Time: O(n), Space: O(1) - fixed charset size
# Unicode handling: Counter works with Unicode by default

from collections import Counter

class Solution:
    # Basic approach
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        count = {}
        
        for char in s:
            count[char] = count.get(char, 0) + 1
        
        for char in t:
            if char not in count:
                return False
            count[char] -= 1
            if count[char] == 0:
                del count[char]
        
        return len(count) == 0
    
    # Using Counter (handles Unicode automatically)
    def isAnagramCounter(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t)

# Test
if __name__ == "__main__":
    sol = Solution()
    print(sol.isAnagram("anagram", "nagaram"))  # True
    print(sol.isAnagram("rat", "car"))          # False
    print(sol.isAnagramCounter("测试", "试测"))  # True (Unicode)

File: 11-valid-anagram/solution.ts
// Valid Anagram
// Time: O(n), Space: O(1) - fixed charset size
// Unicode handling: Map works with Unicode by default

class Solution {
    // Basic approach
    isAnagram(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        
        const count = new Map<string, number>();
        
        for (const char of s) {
            count.set(char, (count.get(char) || 0) + 1);
        }
        
        for (const char of t) {
            if (!count.has(char)) {
                return false;
            }
            const newCount = count.get(char)! - 1;
            if (newCount === 0) {
                count.delete(char);
            } else {
                count.set(char, newCount);
            }
        }
        
        return count.size === 0;
    }
    
    // Using object-based counting
    isAnagramObject(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        
        const count: Record<string, number> = {};
        for (const char of s) {
            count[char] = (count[char] || 0) + 1;
        }
        
        for (const char of t) {
            if (!count[char]) {
                return false;
            }
            count[char]--;
        }
        
        return Object.values(count).every(val => val === 0);
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.isAnagram("anagram", "nagaram"));  // true
    console.log(sol.isAnagram("rat", "car"));          // false
    console.log(sol.isAnagram("测试", "试测"));         // true (Unicode)
}

export default Solution;
